docs/IPC.md

```diff
@@ -35,3 +76,215 @@ Each subscriber's library:
+
+## `socketdriver`
+
+The primary `Driver` implemented in eve-os is the [socketdriver](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub/socketdriver), which uses:
+
+* persistence via files, one directory per table, one file per key-value entry.
+* notification via Unix-domain sockets, one socket per table.
+
+The socketdriver stores its data in a specific directory. The root directory upon which socketdriver operates is passed to the socketdriver
+instance upon creation. In the case of normal eve-os operation, that defaults to `/`, but it can be changed upon initialization.
+All other directories are subsidiary to the root directory.
+
+The specific subdirectory used inside the root directory is determined based on the options [here](https://github.com/lf-edge/eve/blob/6160d0e96c72a1954db2a8bdfd99c2fec1972341/pkg/pillar/pubsub/socketdriver/driver.go#L88-L99).
+
+The directory to use and the type of file depends on 2 options:
+
+* `persistent`: whether or not this table should persist. In both cases, this means the data is written to a directory. In the `persistent` case, the directory is one set to persist beyond reboots; in the non-persistent case, it is in `/run` and reset between reboots.
+* `publishToDir`: whether or not this table should publish its data to a directory.
+
+It also includes the "name" of the publication, where `<name>` is calculated:
+
+* global table: `global`
+* `agentScope == ""`: `<agentName>/<topic>`
+* otherwise: `<agentName>/<agentScope>/<topic>`
+
+The above combine to create the location:
+
+1. persistent AND publish: `/persist/config/<name>`
+1. persistent AND NOT publish: `/persist/<name>`
+1. NOT persistent AND publish: `/persist/global/<name>`
+1. !persistent AND !publish: `/var/run/<name>`
+
+Examples:
+
+|---|---|---|---|---|---|
+|persistent|publish|agentScope|agentName|topic|directory|
+|Y|Y|tester|configmgr|inputs|`/persist/config/tester/configmgr/inputs/`|
+|Y|N|tester|configmgr|inputs|`/persist/tester/configmgr/inputs/`|
+|Y|N||configmgr|inputs|`/persist/configmgr/inputs/`|
+|N|N|||inputs|`/run/global/inputs`|
+
+Once the directory has been determined, if the `DriverPublisher` should be publishing on a socket, which it does in every case except
+for `publishToDir`, it determines the filename for the Unix domain socket. This is:
+
+```shell
+/var/run/<name>.sock
+```
+
+This uses the same `<name>` algorithm as above.
+
+This socket is unique to this table; each table has its own socket.
+
+The specific implementations of the `DriverPublisher` interface and `DriverSubscriber` interface, for socketdriver, are, respectively,
+[socketdriver.Publisher](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub/socketdriver#Publisher) and [socketdriver.Subscriber](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar@v0.0.0-20220603153046-23f5ce4eb5ee/pubsub/socketdriver#Subscriber).
+
+### `socketdriver.Publisher` Updates
```

Comment:
In general the reminder of this document is difficult to digest for me. I tried to comment on the main points of confusion I had. I probably will be adding more as I progress in understanding.

Reply:
I know it is. This could be simpler. It was a _lot_ harder when all of this functionality was mixed together, before we separated it into a driver. Now, at least, you can think of the higher-level as "take an object, store it in memory, give it to something else that handles persistence and notifications and such", and the lower-level as "given raw data, persist it if asked, and publish it to any subscribers".

Hopefully with the updates based on your good comments above, it will make a little more sense. :-)